exp/controllers/awsmachinepool_controller.go

```diff
@@ -497,29 +520,50 @@ func (r *AWSMachinePoolReconciler) findASG(machinePoolScope *scope.MachinePoolSc
 	if !scope.ReplicasExternallyManaged(machinePoolScope.MachinePool) {
 		if machinePoolScope.MachinePool.Spec.Replicas != nil {
-			if existingASG.DesiredCapacity == nil || *machinePoolScope.MachinePool.Spec.Replicas != *existingASG.DesiredCapacity {
+			if existingASG.DesiredCapacity == nil {
+				machinePoolScope.Info("DesiredCapacity unset, diff detected", "incoming", *machinePoolScope.MachinePool.Spec.Replicas, "existing", existingASG.DesiredCapacity)
+				return true
+			}
+			if *machinePoolScope.MachinePool.Spec.Replicas != *existingASG.DesiredCapacity {
+				machinePoolScope.Info("DesiredCapacity diff detected", "incoming", *machinePoolScope.MachinePool.Spec.Replicas, "existing", *existingASG.DesiredCapacity)
 				return true
 			}
 		} else if existingASG.DesiredCapacity != nil {
+			machinePoolScope.Info("Replicas nil, DesiredCapacity diff detected", "incoming", machinePoolScope.MachinePool.Spec.Replicas, "existing", *existingASG.DesiredCapacity)
 			return true
 		}
 	}
 
 	if machinePoolScope.AWSMachinePool.Spec.MaxSize != existingASG.MaxSize {
+		machinePoolScope.Info("MinSize diff detected", "incoming", machinePoolScope.AWSMachinePool.Spec.MaxSize, "existing", existingASG.MaxSize)
 		return true
 	}
 
 	if machinePoolScope.AWSMachinePool.Spec.MinSize != existingASG.MinSize {
+		machinePoolScope.Info("MinSize diff detected", "incoming", machinePoolScope.AWSMachinePool.Spec.MinSize, "existing", existingASG.MinSize)
 		return true
 	}
 
 	if machinePoolScope.AWSMachinePool.Spec.CapacityRebalance != existingASG.CapacityRebalance {
+		machinePoolScope.Info("CapacityRebalance diff detected", "incoming", machinePoolScope.AWSMachinePool.Spec.CapacityRebalance, "existing", existingASG.CapacityRebalance)
 		return true
 	}
 
-	if !cmp.Equal(machinePoolScope.AWSMachinePool.Spec.MixedInstancesPolicy, existingASG.MixedInstancesPolicy) {
-		machinePoolScope.Info("got a mixed diff here", "incoming", machinePoolScope.AWSMachinePool.Spec.MixedInstancesPolicy, "existing", existingASG.MixedInstancesPolicy)
-		return true
+	{
+		mixedInstancesPolicy := machinePoolScope.AWSMachinePool.Spec.MixedInstancesPolicy
+
+		// InstancesDistribution is optional, and the default values come from AWS, so
+		// they are not set by the AWSMachinePool defaulting webhook. If InstancesDistribution is
+		// not set, we use the AWS values for the purpose of comparison.
+		if mixedInstancesPolicy != nil && mixedInstancesPolicy.InstancesDistribution == nil {
+			mixedInstancesPolicy = machinePoolScope.AWSMachinePool.Spec.MixedInstancesPolicy.DeepCopy()
+			mixedInstancesPolicy.InstancesDistribution = existingASG.MixedInstancesPolicy.InstancesDistribution
+		}
+
+		if !cmp.Equal(mixedInstancesPolicy, existingASG.MixedInstancesPolicy) {
+			machinePoolScope.Info("MixedInstancesPolicy diff detected", "incoming", spew.Sprintf("%#v", mixedInstancesPolicy), "existing", spew.Sprintf("%#v", existingASG.MixedInstancesPolicy))
```

Comment:
We never use spew in the logs. I'm pretty sure this would look super ugly. :D Can you print individual values instead or somehow make a prettier output?

Reply:
Not without expanding the footprint of the code there significantly. `MixedInstancesPolicy` is optional (could be nil on either object) and has further optional nil-able inner struct fields, and golang lacks a safe navigator operator :(

The output is the struct object. `%#v` shows the struct + field names, so it's human readable.

Reply:
I think our goal is to help the user understand the diff, if one occurs. For representing the diff, I think https://pkg.go.dev/github.com/google/go-cmp/cmp#Diff is very readable. Had you already considered it?

For example, this output

```console
2023/05/16 14:58:17 (-existing, +actual)   &v1beta2.MixedInstancesPolicy{
  	InstancesDistribution: &v1beta2.InstancesDistribution{
- 		OnDemandAllocationStrategy:          "",
+ 		OnDemandAllocationStrategy:          "prioritized",
- 		SpotAllocationStrategy:              "lowest-price",
+ 		SpotAllocationStrategy:              "",
  		OnDemandBaseCapacity:                nil,
  		OnDemandPercentageAboveBaseCapacity: nil,
  	},
  	Overrides: nil,
  }
````

is the result of the example

```go
package main

import (
	"log"

	"github.com/google/go-cmp/cmp"

	expinfrav1 "sigs.k8s.io/cluster-api-provider-aws/v2/exp/api/v1beta2"
)

func main() {
	existing := &expinfrav1.MixedInstancesPolicy{
		InstancesDistribution: &expinfrav1.InstancesDistribution{
			SpotAllocationStrategy: expinfrav1.SpotAllocationStrategyLowestPrice,
		},
		Overrides: nil,
	}

	actual := &expinfrav1.MixedInstancesPolicy{
		InstancesDistribution: &expinfrav1.InstancesDistribution{
			OnDemandAllocationStrategy: expinfrav1.OnDemandAllocationStrategyPrioritized,
		},
		Overrides: nil,
	}
	diff := cmp.Diff(existing, actual)
	log.Println("(-existing, +actual)", diff)
}
```

However, the output is readable because it's multiline, and our logs are structured, so I think the newlines will be escaped and everything fit into a single line, which will be very hard to read.

Have you considered logging that the diff occurred, but writing the diff itself to an Event?

Reply:
> I think our goal is to help the user understand the diff, if one occurs. For representing the diff, I think https://pkg.go.dev/github.com/google/go-cmp/cmp#Diff is very readable. Had you already considered it?

I'm not very familiar with that library, so no. But that functionality looks like it could produce the output we want. 

@dlipovetsky what if we use inspiration from your previous suggestion about how to refactor the  `mixedInstancesPolicy` to take a different approach, using `cmp.Diff` to detect the differences in the ASG, and then use that output for the event?


```
// detectASGDrift compares incoming AWSMachinePool and compares against existing ASG.
func detectASGDrift(machinePoolScope *scope.MachinePoolScope, existingASG *expinfrav1.AutoScalingGroup) string {
	detectedMachinePoolSpec := machinePoolScope.MachinePool.Spec.DeepCopy()

	if !scope.ReplicasExternallyManaged(machinePoolScope.MachinePool) {
		detectedMachinePoolSpec.Replicas = existingASG.DesiredCapacity
	}
	if diff := cmp.Diff(machinePoolScope.MachinePool.Spec, *detectedMachinePoolSpec); diff != "" {
		return diff
	}

	detectedAWSMachinePoolSpec := machinePoolScope.AWSMachinePool.Spec.DeepCopy()
	detectedAWSMachinePoolSpec.MaxSize = existingASG.MaxSize
	detectedAWSMachinePoolSpec.MinSize = existingASG.MinSize
	detectedAWSMachinePoolSpec.CapacityRebalance = existingASG.CapacityRebalance
	{
		mixedInstancesPolicy := machinePoolScope.AWSMachinePool.Spec.MixedInstancesPolicy
		// InstancesDistribution is optional, and the default values come from AWS, so
		// they are not set by the AWSMachinePool defaulting webhook. If InstancesDistribution is
		// not set, we use the AWS values for the purpose of comparison.
		if mixedInstancesPolicy != nil && mixedInstancesPolicy.InstancesDistribution == nil {
			mixedInstancesPolicy = machinePoolScope.AWSMachinePool.Spec.MixedInstancesPolicy.DeepCopy()
			mixedInstancesPolicy.InstancesDistribution = existingASG.MixedInstancesPolicy.InstancesDistribution
		}

		if !cmp.Equal(mixedInstancesPolicy, existingASG.MixedInstancesPolicy) {
			detectedAWSMachinePoolSpec.MixedInstancesPolicy = existingASG.MixedInstancesPolicy
		}
	}

	return cmp.Diff(machinePoolScope.AWSMachinePool.Spec, *detectedAWSMachinePoolSpec)
}
```

And then in `updatePool` use that:

```
	asgDiff := detectASGDrift(machinePoolScope, existingASG)
	if asgDiff != "" {
		r.Recorder.Eventf(machinePoolScope.AWSMachinePool, corev1.EventTypeNormal, "ASGDiffDetected", asgDiff)
	}
	if asgDiff != "" || subnetChanges {
		machinePoolScope.Info("updating AutoScalingGroup")
...
```

If prefer that approach, I can update this PR to do that and run some tests.

Reply:
I like it!